src/SkeletalTrapezoidationGraph.cpp

```diff
@@ -0,0 +1,323 @@
+#include "SkeletalTrapezoidationGraph.h"
+#include <unordered_map>
+
+#include "utils/linearAlg2D.h"
+
+namespace arachne
+{
+
+void SkeletalTrapezoidationGraph::fixNodeDuplication()
+{
+    for (auto node_it = nodes.begin(); node_it != nodes.end();)
+    {
+        node_t* replacing_node = nullptr;
+        for (edge_t* outgoing = node_it->some_edge; outgoing != node_it->some_edge; outgoing = outgoing->twin->next)
+        {
+            assert(outgoing);
+            if (outgoing->from != &*node_it)
+            {
+                replacing_node = outgoing->from;
+            }
+            if (outgoing->twin->to != &*node_it)
+            {
+                replacing_node = outgoing->twin->to;
+            }
+        }
+        if (replacing_node)
+        {
+            for (edge_t* outgoing = node_it->some_edge; outgoing != node_it->some_edge; outgoing = outgoing->twin->next)
+            {
+                outgoing->twin->to = replacing_node;
+                outgoing->from = replacing_node;
+            }
+            node_it = nodes.erase(node_it);
+        }
+        else
+        {
+            ++node_it;
+        }
+    }
+}
+    
+void SkeletalTrapezoidationGraph::collapseSmallEdges(coord_t snap_dist)
+{
+    std::unordered_map<edge_t*, std::list<edge_t>::iterator> edge_locator;
+    std::unordered_map<node_t*, std::list<node_t>::iterator> node_locator;
+    
+    for (auto edge_it = edges.begin(); edge_it != edges.end(); ++edge_it)
+    {
+        edge_locator.emplace(&*edge_it, edge_it);
+    }
+    
+    for (auto node_it = nodes.begin(); node_it != nodes.end(); ++node_it)
+    {
+        node_locator.emplace(&*node_it, node_it);
+    }
+    
+    auto safelyRemoveEdge = [this, &edge_locator](edge_t* to_be_removed, std::list<edge_t>::iterator& current_edge_it, bool& edge_it_is_updated)
```

Comment:
Consider making this a normal function of the `HalfEdgeGraph` rather than the anonymous function here. See my previous comment here: https://github.com/Ultimaker/CuraEngine/pull/1210#discussion_r392095565

Reply:
It removes the edge from the unordered map `edge_locator`, not from the graph.

Or do you want these maps to be in the edge as well? They're only used in `collapseSmallEdges`, so I don't think that would make the code more clear in this particular case.

Reply:
I agree, that wouldn't be more readable and wouldn't matter for performance either, so let's not do that then.